subprojects/core/src/main/java/org/gradle/execution/TaskGraphBuildExecutionAction.java

```diff
@@ -0,0 +1,190 @@
+            }
+        }
+    }
+
+    private static class NodesGraph implements DirectedGraph<TaskInfo, TaskInfo> {
+
+        @Override
+        public void getNodeValues(TaskInfo node, Collection<? super TaskInfo> values, Collection<? super TaskInfo> connectedNodes) {
+            values.add(node);
+            connectedNodes.addAll(node.getDependencies());
+        }
+    }
+
+    private static Stream<TaskInfo> extractTaskNodes(Collection<Node> collection, DependencyType type) {
+        return collection.stream()
+            .filter(node -> node instanceof TaskNode && !node.isDoNotIncludeInPlan())
+            .map(taskNode -> new DefaultTaskInfo((TaskNode) taskNode, type));
+    }
+
+    private enum DependencyType {
+        REGULAR,
+        FINALIZING
+    }
+
+    private interface TaskInfo {
+        String getId();
+
+        String getDescription();
+
+        Collection<TaskInfo> getDependencies();
+    }
+
+    private static class RootNode implements TaskInfo {
+        private final Collection<Node> entryNodes;
+        private final String invocationInfo;
+
+        public RootNode(Collection<Node> entryNodes, String invocationInfo) {
+            this.entryNodes = entryNodes;
+            this.invocationInfo = invocationInfo;
+        }
+
+        @Override
+        public String getId() {
+            return "Tasks graph for: " + invocationInfo;
+        }
+
+        @Override
+        public String getDescription() {
+            return "";
+        }
+
+        @Override
+        public Collection<TaskInfo> getDependencies() {
+            return extractTaskNodes(entryNodes, DependencyType.REGULAR)
+                .collect(Collectors.toList());
+        }
+    }
+
+    private static class DefaultTaskInfo implements TaskInfo {
+        private final TaskNode node;
+        private final DependencyType type;
+
+        DefaultTaskInfo(TaskNode node, DependencyType type) {
+            this.node = node;
+            this.type = type;
+        }
+
+        @Override
+        public String getId() {
+            return node.toString();
+        }
+
+        @Override
+        public String getDescription() {
+            StringBuilder description = new StringBuilder();
+            description.append("(");
+            description.append(node.getTask().getTaskIdentity().getTaskType().getCanonicalName());
+            if (type.equals(DependencyType.FINALIZING)) {
+                description.append(", finalizer");
+            }
+            if (!node.getTask().getEnabled()) {
+                description.append(", disabled");
+            }
+            description.append(")");
+            return description.toString();
+        }
+
+        @Override
+        public Collection<TaskInfo> getDependencies() {
+            return Streams.concat(
+                extractTaskNodes(node.getDependencySuccessors(), DependencyType.REGULAR),
+                extractTaskNodes(node.getFinalizers(), DependencyType.FINALIZING)
```

Comment:
💭 We should ask the Execution team if there is anything we might be missing here.